### PR TITLE
Uniformization of internal identifiers generation

### PR DESCRIPTION
**Pull request by Stevendeo**
This PR centralizes how internal identifiers are created. This work is a first step before a simpler task, which is to replace the '!' by '.' in name identifiers.

### CODE REVIEW

File: src/lib/frontend/typechecker.ml
Comment (Halbaroth) [inline src/lib/frontend/typechecker.ml]: I suggest to use `InternalId` instead of `InternalNameId`.

File: src/lib/structures/symbols.mli
Comment (Halbaroth) [inline src/lib/structures/symbols.mli]: Is it necessary to expose the function `make_as_fresh`?

File: src/lib/structures/symbols.ml
Comment (hra687261) [inline src/lib/structures/symbols.ml]:
```suggestion
    in len_s > len_pre && aux 0
```
Assuming that names as empty strings don't occur

File: src/lib/structures/symbols.mli
Comment (hra687261) [inline src/lib/structures/symbols.mli]: Suggestion, just for consistency:
`fresh_name` -> `fresh_interna_name`
`is_internal_name` -> `is_fresh_internal_name`
`is_skolem` -> `is_fresh_skolem`

**Review by hra687261 — APPROVED** [review]
LGTM modulo a couple of minor details
If my memory serves me right, these fresh names are used for comparisons in some cases, out of prudence, we should probably run some benchmarks, before merging.

File: src/lib/structures/symbols.ml
Comment (Stevendeo) [reply src/lib/structures/symbols.ml]: `||` is a valid SMTLib2 identifier corresponding to the empty string.

File: src/lib/structures/symbols.mli
Comment (Stevendeo) [reply src/lib/structures/symbols.mli]: Unfortunately, yes... https://github.com/OCamlPro/alt-ergo/pull/905/files#diff-89722ab84fe15db7d6cd1e3a8533c9d5e0660599698ae64313faae9cf6f23730L1720

Comment (Halbaroth) [reply src/lib/structures/symbols.mli]: You only need `make_as_fresh_skolem` here ;)

File: src/lib/structures/symbols.ml
Comment (Halbaroth) [reply src/lib/structures/symbols.ml]: Oh gosh... the SMT-LIB accepts too much things as identifiers...

File: src/lib/structures/symbols.mli
Comment (Stevendeo) [reply src/lib/structures/symbols.mli]: Oh okay I get your point; if I remove the `make_as_fresh`, I can also remove the `reset_fresh_counter` and hide the whole module.

Comment (bclement-ocp) [reply src/lib/structures/symbols.mli]: We may need to keep `reset_fresh_counter` for alt-ergo-fuzz, @hra687261 can you confirm?

File: src/lib/structures/symbols.ml
Comment (bclement-ocp) [reply src/lib/structures/symbols.ml]: I think this is a reimplementation of `Compat.String.starts_with ~prefix:S.prefix`?

File: src/lib/structures/expr.ml
Comment (bclement-ocp) [inline src/lib/structures/expr.ml:1229]: I think that `Sy.fresh` used to add an extra `__` prefix (with the comment "(* garder le suffixe "__" car cela influence l'ordre *)") but it looks like `Sy.fresh_skolem` doesn't, so unless I missed something we used to get `!?__@sko` and we now get `!?@sko` here.

**Review by bclement-ocp — CHANGES_REQUESTED** [review]
Agree with @hra687261 — this needs to be tested at least on ae-format and benchtop-tests. Marking as "change requested" so that we don't merge by mistake yet.

(Note that preserving the exact names of the symbols don't matter; *if* there is no change in the problems solved — but the comments hint that there may be some)

File: src/lib/structures/symbols.mli
Comment (Stevendeo) [reply src/lib/structures/symbols.mli]: > We may need to keep `reset_fresh_counter` for alt-ergo-fuzz, @hra687261 can you confirm?

I meant hide the module by wrapping the reset_fresh_cpt in a specific function (which is what I did in the last version of the PR)

File: src/lib/structures/symbols.ml
Comment (Stevendeo) [reply src/lib/structures/symbols.ml]: > I think this is a reimplementation of `Compat.String.starts_with ~prefix:S.prefix`?

Oh yes indeed

File: src/lib/structures/expr.ml
Comment (hra687261) [reply src/lib/structures/expr.ml:1229]: As it is outdated I can't reply to the right comment, but `reset_id_builders` will do instead of `reset_fresh_counter`, thanks for noticing that @bclement-ocp !

Comment (Stevendeo) [reply src/lib/structures/expr.ml:1229]: > I think that `Sy.fresh` used to add an extra `__` prefix (with the comment "(* garder le suffixe "__" car cela influence l'ordre *)") but it looks like `Sy.fresh_skolem` doesn't, so unless I missed something we used to get `!?__@sko` and we now get `!?@sko` here.

That makes me think this whole check system is unsound, because an identifier can start with `.?` if is has been defined quoted.

Comment (bclement-ocp) [reply src/lib/structures/expr.ml:1229]: This is why we want to replace `!` with `.`! Identifiers starting with `.` or `@` are reserved for solver usage and cannot be used in scripts, dolmen ensures this (or will in the next release, see Gbury/dolmen#185 ).

Comment (bclement-ocp) [reply src/lib/structures/expr.ml:1229]: Ah, sorry, I misread your remark — indeed, according to the SMT-LIB it can, but I think after Gbury/dolmen#185 identifiers starting with `.?` are rejected even in quoted form (which is incorrect from a strict reading of the standard).

Comment (Stevendeo) [reply src/lib/structures/expr.ml:1229]: > Ah, sorry, I misread your remark — indeed, according to the SMT-LIB it can, but I think after [Gbury/dolmen#185](https://github.com/Gbury/dolmen/issues/185) identifiers starting with `.?` are rejected even in quoted form (which is incorrect from a strict reading of the standard).

You did not, I miswrote it and edited, you might have seen the old version :)

**Review by bclement-ocp — COMMENTED** [review]
I believe these changes are correct and should not change the identifiers we generate, so there should be zero performance impact here, but please run benchmarks on `ae-format` and `benchtop-tests` to confirm.

**Comment by bclement-ocp** [comment]
I am not sure why you re-requested a review from me — I think we are good and just waiting on benchmark results, or is there a specific part you want me to look at?

**Comment by Stevendeo** [comment]
> I am not sure why you re-requested a review from me — I think we are good and just waiting on benchmark results, or is there a specific part you want me to look at?

Still me discovering github, sorry about that

**Comment by Stevendeo** [comment]
No regressions (and even 1 improvement), we can merge!

**Comment by Stevendeo** [comment]
@bclement-ocp Can we merge now?

**Review by bclement-ocp — APPROVED** [review]

**Comment by bclement-ocp** [comment]
I think so — sorry I got confused somehow, thought I had approved already.

**Pull request merged**
